### PR TITLE
Rework native surfaces for flexibility and simplicity

### PR DESCRIPTION
**Pull request by mrobinson**
Currently, the Linux native surface can be either a memory-buffer
surface or a Pixmap, chosen at runtime. Since the memory-buffer surface
is platform-independent, we push this configurability to the top-level
native surface API. This will allow having multiple surface types
per-platform, which is useful for both adding Wayland support and
adding configurable surface transports for performance tuning. This
also simplifies the surface interface a great deal.

The other reason for this change is it prepares the rust-layers
implementation for handling more of the glue between Azure, skia, and
Servo, so that the first two do not need intimate details of the Servo
compositor. This will also make it easier to rework how DrawTargets in
Azure interact with rust-layers native surfaces.

### CODE REVIEW

**Comment by mrobinson** [comment]
@glennw r? This is a generalization of your earlier change.

File: src/platform/surface.rs [hunk 1]
```diff
@@ -82,6 +180,57 @@ pub trait NativeSurfaceMethods {
...
+        native_surface_method_mut!(self mark_wont_leak)
+    }
+}
+
+#[deriving(Decodable, Encodable)]
+pub struct MemoryBufferNativeSurface {
+    bytes: Vec<u8>,
+}
+
+impl Drop for MemoryBufferNativeSurface {
```
Comment (glennw) [inline src/platform/surface.rs]: Can we remove this or is it needed for some reason?

Comment (mrobinson) [inline src/platform/surface.rs]: I think you added it for headless support. I'm not sure if it's still necessary or not.

**Comment by glennw** [comment]
Seems like a straighforward generalization like you mentioned. I don't have my mac/android builds here on this laptop, so haven't verified them but the changes look good, just the one comment / question above.

File: src/platform/surface.rs (hunk 1 above)
Comment (mrobinson) [inline src/platform/surface.rs]: It seems that this Drop implementation can be omitted. Everything compiles fine without it.

**Comment by glennw** [comment]
r+

**Pull request merged**